packages/expo-permissions/ios/EXPermissions/EXPermissions.m

```diff
@@ -28,266 +27,209 @@ @implementation EXPermissions
-      return [requesterClass permissionsWithModuleRegistry:_moduleRegistry];
-    }
+  return [self getPermissionUsingRequester:_requesters[type]];
+}
 
-    return [requesterClass permissions];
+- (NSDictionary *)getPermissionUsingRequester:(id<UMPermissionsRequester>)requester
+{
+  if (requester) {
+    return [EXPermissions parsePermissionFromRequester:[requester getPermissions]];
   }
-
   return nil;
 }
 
 // shorthand method that checks both global and per-experience permission
-- (BOOL)hasGrantedPermission:(NSString *)permissionType
+- (BOOL)hasGrantedPermissionUsingRequesterClass:(Class)requesterClass
 {
-  NSDictionary *permissions = [self getPermissionsForResource:permissionType];
-
+  NSDictionary *permissions = [self getPermissionUsingRequesterClass:requesterClass];
   if (!permissions) {
-    UMLogWarn(@"Permission with type '%@' not found.", permissionType);
+    UMLogWarn(@"Permission requester '%@' not found.", NSStringFromClass(requesterClass));
     return false;
   }
   
-  return [permissions[@"status"] isEqualToString:@"granted"];
+  return [permissions[EXStatusKey] isEqualToString:@"granted"];
 }
 
-- (void)askForPermission:(NSString *)permissionType
-              withResult:(void (^)(NSDictionary *))onResult
-            withRejecter:(UMPromiseRejectBlock)reject
+- (void)askForPermissionUsingRequesterClass:(Class)requesterClass
+                                 withResult:(UMPromiseResolveBlock)onResult
+                               withRejecter:(UMPromiseRejectBlock)reject
 {
-  return [self askForPermissions:@[permissionType]
-                     withResults:^(NSArray<NSDictionary *> *results) {
-                       onResult(results[0]);
-                     }
-                    withRejecter:reject];
+  NSMutableDictionary *permission = [[self getPermissionUsingRequesterClass:requesterClass] mutableCopy];
+  
+  // permission type not found - reject immediately
+  if (permission == nil) {
+    return reject(@"E_PERMISSIONS_UNKNOWN", [NSString stringWithFormat:@"Unrecognized requester: %@", NSStringFromClass(requesterClass)], nil);
+  }
+  
+  BOOL isGranted = [EXPermissions statusForPermission:permission] == UMPermissionStatusGranted;
+  permission[@"granted"] = @(isGranted);
+  
+  if (isGranted) {
+    return onResult(permission);
+  }
+  
+  id<UMPermissionsRequester> requester = [self getPermissionRequesterForClass:requesterClass];
+  [self askForGlobalPermissionUsingRequester:requester withResolver:onResult withRejecter:reject];
 }
-
-- (void)askForPermissions:(NSArray<NSString *> *)permissionsTypes
-              withResults:(void (^)(NSArray<NSDictionary *> *))onResults
-             withRejecter:(UMPromiseRejectBlock)reject
+   
+- (void)askForGlobalPermissionUsingRequesterClass:(Class)requesterClass
+                                     withResolver:(UMPromiseResolveBlock)resolver
+                                     withRejecter:(UMPromiseRejectBlock)reject
```

Comment:
same comment here, and in the rest of the methods added to this module -- should be resolver:resolve and rejecter:reject